src/MobileRedirectManager.php

```diff
@@ -0,0 +1,183 @@
+<?php
+/**
+ * Class MobileRedirect.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AMP_Options_Manager;
+
+/**
+ * Centralized manager to handle mobile redirection of a page.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class MobileRedirectManager {
+
+	/**
+	 * The name of the cookie that persists the user's preference for viewing the non-AMP version of a page when on mobile.
+	 *
+	 * @var string
+	 */
+	const DISABLED_COOKIE_NAME = 'amp_mobile_redirect_disabled';
+
+	/**
+	 * Query parameter to indicate that the page in question should not be served as AMP.
+	 *
+	 * @var string
+	 */
+	const NO_AMP_QUERY_VAR = 'noamp';
+
+	/**
+	 * Get whether mobile redirection is enabled or not.
+	 *
+	 * @return bool If JS redirection is disabled, only the status of the mobile redirection option is returned.
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
+		$option_enabled = AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT );
+
+		return self::should_redirect_via_js() ? $option_enabled : $option_enabled && self::is_mobile();
+	}
+
+	/**
+	 * Determine if the current request is from a mobile device.
+	 *
+	 * @return bool True if current request if from a mobile device, otherwise false.
+	 */
+	public static function is_mobile() {
+		$pre_is_mobile = apply_filters( 'amp_pre_is_mobile', false );
+
+		if ( true === $pre_is_mobile ) {
+			return (bool) $pre_is_mobile;
+		}
```

Comment:
Should this be here? There'd be no way to replicate this in JS.

Reply:
The reasoning here is that the user may have already determined that the current request is from a mobile device, maybe from a reverse proxy or cookie, for example. If that's the case then the server-side logic below would become redundant.